packages/service-clients/end-to-end-tests/azure-client/src/test/multiprocess/presenceTest.spec.ts

```diff
@@ -0,0 +1,148 @@
+ * Currently our E2E tests are limited to running multiple clients on a single process which does not effectively
+ * simulate real-world production scenarios where clients are usually running on different machines.
+ *
+ * The pattern demonstrated in this test suite is as follows:
+ *
+ * This main test file acts as the 'Orchestrator'. The orchestrator's job includes:
+ * - Fork child processes to simulate multiple Fluid clients
+ * - Send command messages to child clients to perform specific Fluid actions.
+ * - Receive response messages from child clients to verify expected behavior.
+ * - Clean up child processes after each test.
+ *
+ * The child processes are located in the `childClient.ts` file. Each child process simulates a Fluid client.
+ *
+ * The child client's job includes:
+ * - Create/Get + connect to Fluid container.
+ * - Listen for command messages from the orchestrator.
+ * - Perform the requested action.
+ * - Send response messages including any relevant data back to the orchestrator to verify expected behavior.
+ *
+ * This particular test suite tests the following E2E functionality for Presence:
+ * - Announce 'attendeeJoined' when remote client joins session.
+ * - Announce 'attendeeDisconnected' when remote client disconnects.
+ */
+describe(`Presence with AzureClient`, () => {
+	const numClients = 5; // Set the total number of Fluid clients to create
+	assert(numClients > 1, "Must have at least two clients");
+	let children: ChildProcess[] = [];
+	const durationMs = 10_000;
+
+	const afterCleanUp: (() => void)[] = [];
+
+	// After each test, kill each child process and call any cleanup functions that were registered
+	afterEach(async () => {
+		for (const child of children) {
+			child.kill();
+		}
+		children = [];
+		for (const cleanUp of afterCleanUp) {
+			cleanUp();
+		}
+		afterCleanUp.length = 0;
+	});
+
+	beforeEach("setup", async () => {
+		// Create inital child process
+		let containerId: string | undefined;
+
+		// Fork child processes
+		for (let i = 0; i < numClients; i++) {
+			const child = fork("./lib/test/multiprocess/childClient.js", [
+				`child${i}` /* identifier passed to child process */,
+			]);
+			children.push(child);
+		}
+
+		// Send connect command to each child process
+		for (const [index, child] of children.entries()) {
+			const user = { id: `test-user-id-${index}`, name: `test-user-name-${index}` };
+			const message: MessageToChild = { command: "connect", containerId, user };
+			child.send(message);
+			/*
+			 The initial child process will create the container,
+			 so we must wait to receive the containerId so future child clients can use it
+			*/
+			if (index === 0) {
+				await timeoutPromise(
+					(resolve) => {
+						child.once("message", (msg: MessageFromChild) => {
+							if (msg.event === "ready") {
+								containerId = msg.containerId;
+								resolve();
+							}
+						});
+					},
```

Comment:
You can't use `once` here and also filter the events just ignoring non-ready.
You could reject on anything else
```suggestion
				await timeoutPromise(
					(resolve, reject) => {
						child.once("message", (msg: MessageFromChild) => {
							if (msg.event === "ready") {
								containerId = msg.containerId;
								resolve();
							} else {
								reject(new Error(`Non-ready message from child0: ${JSON.stringify(msg)}`));
							}
						});
					},
```
Additionally, at least the inner promise on listener should be setup _before_ any message is sent; so, there is no whole where a message could be missed. It would probably be fine to create the timeoutPromise before and just await in the if. (Maybe local is a `const postWait = (index === 0) ? ... : undefined;`, then you can just await if defined instead of two ifs for index 0.